Baseline set of fixes for release

agibsonccc opened the pull request:
## What changes were proposed in this pull request?

(Please fill in changes proposed in this fix)

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)

## Quick checklist

The following checklist helps ensure your PR is complete:

- [ ] Eclipse Contributor Agreement signed, and signed commits - see [IP Requirements](https://deeplearning4j.konduit.ai/multi-project/how-to-guides/contribute/eclipse-contributors) page for details
- [ ] Reviewed the [Contributing Guidelines](https://github.com/eclipse/deeplearning4j/blob/master/CONTRIBUTING.md) and followed the steps within.
- [ ] Created tests for any significant new code additions.
- [ ] Relevant tests for your changes are passing.

agibsonccc committed:
Fix ndarray compilation
Fix left over test issue
Add hack that configures surefire to use a wrapper that allows us to debug tests

agibsonccc committed:
Fix ndarray compilation
Fix left over test issue
Add hack that configures surefire to use a wrapper that allows us to debug tests

agibsonccc committed:
Add new build scripts
Fix up deallocation race conditions with constants (removes random deallocation of constants)

Bring back some deallocations

agibsonccc committed:
Add deallocation  and allocation logging behind a flag to allow better debugging of stack traces
when SD_GCC_FUNCTRACE is turned on and the right environment is set.
Remove old dependencies from tests

agibsonccc committed:
Fix up transpose op bugs. (deallocations and shape buffer related)
Ensure thread safety on deallocations in pointerwrapper,opaquebuffer
Fix up uniform dealocation and result propagation.
Remove commented code.
Remove old traceNew method
Removes old clionide exclusion in type_boilerplate.h
Remove unused scalar.cu
Misc fixes for scalar related shape descriptor creation
Update java build script
Remove some unused c++ tests
Improve debug statements in DeclarableOp.cpp
Add new FuncTracce developer doc explaining the functionality.

agibsonccc committed:
Clean up string assign support
More deleted code

agibsonccc committed:
Add proper thread sanitizer support
Add flag on environment to enable/disable delete special/primary buffers
Misc unused code cleanup

agibsonccc committed:
Fix up scalar shape builders support

agibsonccc committed:
WIP:
Empty flags fixes across various ops
Various flag fixes for each op

agibsonccc committed:
Fix dynamic stitch in cuda
Remove various empty specific workarounds across different ops

agibsonccc committed:
Fix conv2d cuda invocation

agibsonccc committed:
Fix cpu build

agibsonccc committed:
Add new result set printing
Fix up solve for cuda

agibsonccc committed:
Remove print statements
Add lup dimensions

agibsonccc committed:
Add databuffer debugging
Add databuffer print for cuda/cpu

agibsonccc committed:
Remove print statements
Add softmax launch dimensions
Fix view write bug with assign and cast

agibsonccc committed:
Fix underlyng tad  cuda issue
Fix softmax

agibsonccc committed:
Fix solve to reach parity with cuda

agibsonccc committed:
Fix array options data type
Add validation for data type creation

agibsonccc committed:
Fix array options data type
Add validation for data type creation

agibsonccc committed:
Ensure a check is performed for empty bufers when validating data types

agibsonccc committed:
Add partitioned tensorflow tests to alleviate cuda resource constraints
Add ending statement to printIndexedBuffer for easier searching when large arrays are output
Remove im2col/col2im configurations in testing scripts

agibsonccc committed:
Remove old assumption about reduce + empty shapes always being 0 length.

agibsonccc committed:
Fix range shape return type
Fix bool data types

agibsonccc committed:
Fix up more range invocations and empty shapes
Fix squeeze and slice empty shapes

agibsonccc committed:
Fix all TF partitions
Port cumsum cpu implementation over to cuda as temp workaround
Fix broadcast_bool edge cases.

agibsonccc committed:
Add more order validation.
Remove prints statements
Automatically return c on 0 rank arrays

agibsonccc committed:
Exclude validation on certain shape descriptor constructors

agibsonccc committed:
Add reshape to flattened params
Misc formatting
Fix up rest of cuda tests
Add more cuda validation to  various kernel launches

agibsonccc committed:
Fix misc reshape issues with 2d biases
Consolidate reflection based classloading for deserialization
and op loading
Remove more legacy helper related code
Fix various databuffer deallocations
Refactor related DifferentialFunctionClassHolder usage
Fix conv1d related undefined behavior/process hangs when a bias is null

agibsonccc committed:
Fix assign on flatten arrays
Reshape would have issues when the array can't be returned as a view.
Incorrect arrays would be created. This affected parameters among other things.
Create new linear_copy to ensure we can bypass calculate output shape with assign and non broadcastable arrays.

agibsonccc committed:
Overhaul workspaces usage in deeplearning4j-nn.
Removes random scope borrow/exit. Changes it to a standardized
input and output workspace enter/exit.

agibsonccc committed:
Update computation graph to achieve parity with new workspace management.

agibsonccc committed:
Fix reverse time series mask

agibsonccc committed:
Add model manager in python  in dl4j contrib.
Add java side model manager for baseline interop and testing.

agibsonccc committed:
Add first bidirectional test
Fix minor model manager issues with computing gradients.

agibsonccc committed:
Fix model manager running models
Fix bidirectional import in keras

agibsonccc committed:
Finish fixing recurrent layer
Finish adding abstractions for the keras model tests
Add concrete implementation based on hand coded test of bidirectional.
Add "well known directory" for keras save_models that java and python use.
Add commented implementation of a base class for tests that use these modules.
Rename things to be clearly tied to keras to avoid confusion with existing samediff tensorflow import tests

Kali-Zoidberg commented:
Hello, I am compiling this with CUDA 12.1 on windows. I will be leaving a couple of comments on areas that led to build failure in libnd4j. Hopefully this will save you some time if the areas haven't been addressed yet.

Kali-Zoidberg commented on libnd4j/include/legacy/cuda/NativeOps.cu:
```diff
@@ -3086,28 +3143,27 @@ sd::Status execCustomOpWithScope(sd::Pointer *extraPointers, sd::graph::GraphSta
-int estimateThreshold(sd::Pointer *extraPointers, sd::Pointer dX, sd::LongType const *dXShapeInfo, int N,
+int estimateThreshold(Pointer *extraPointers, Pointer dX, LongType const *dXShapeInfo, int N,
                       float threshold) {
   THROW_EXCEPTION("estimateThreshold: Not implemented yet");
```
return statement missing - caused ninja build failure

Kali-Zoidberg commented on libnd4j/include/ops/declarable/helpers/cuda/concat.cu:
```diff
@@ -31,38 +31,41 @@
 
 #include <numeric>
 
+#include "../../../../../../../../../../../usr/include/complex.h"
```
Was this line accidentally left in?

Kali-Zoidberg commented on libnd4j/include/ops/declarable/helpers/cuda/dynamic.cu:
```diff
@@ -354,13 +380,13 @@ void dynamicPartitionFunctor(sd::LaunchContext *context, NDArray const *input, N
+static Status _dynamicStitchFunctorBP(std::vector<NDArray *> const &inputs, std::vector<NDArray *> const &indices,
+                                      NDArray const *gradInput, std::vector<NDArray *> &outputList) {
   THROW_EXCEPTION("Not implemented yet");
 }
```
Missing return statement

archenroot commented:
I fetched your pull request and tested and still failing on libnd4j:
```
[ 91%] Building CXX object tests_cpu/layers_tests/CMakeFiles/runtests.dir/LegacyOpsTests.cpp.o
cd /home/zangetsu/proj/public/deeplearning4j/libnd4j/blasbuild/cpu/tests_cpu/layers_tests && /usr/bin/c++ -DCPU_FEATURES=true -DDNNL_ENABLE_CONCURRENT_EXEC=true -DSTACK_LINE_READER_BUFFER_SIZE=1024 -I/home/zangetsu/.javacpp/cache/openblas-0.3.23-1.5.9-linux-x86_64.jar/org/bytedeco/openblas/linux-x86_64/include -I/home/zangetsu/proj/public/deeplearning4j/libnd4j/blasbuild/cpu/cpu_features-src/include -I/home/zangetsu/proj/public/deeplearning4j/libnd4j/blasbuild/cpu/flatbuffers-src/include -I/home/zangetsu/proj/public/deeplearning4j/libnd4j/blasbuild/cpu/include -I/home/zangetsu/proj/public/deeplearning4j/libnd4j/include -I/home/zangetsu/proj/public/deeplearning4j/libnd4j/blasbuild/cpu/tests_cpu/googletest-src/include -I/home/zangetsu/proj/public/deeplearning4j/libnd4j/blasbuild/cpu/tests_cpu/googletest-src -I/home/zangetsu/proj/public/deeplearning4j/libnd4j/tests_cpu/layers_tests/include -isystem /home/zangetsu/proj/public/deeplearning4j/libnd4j/blasbuild/cpu/tests_cpu/googletest-src/googletest/include -isystem /home/zangetsu/proj/public/deeplearning4j/libnd4j/blasbuild/cpu/tests_cpu/googletest-src/googletest -fPIC -DSD_ALL_OPS=true -DBUILD_TESTS=true -fmax-errors=3 -fopenmp -O3 -fPIC -D_RELEASE=true -MD -MT tests_cpu/layers_tests/CMakeFiles/runtests.dir/LegacyOpsTests.cpp.o -MF CMakeFiles/runtests.dir/LegacyOpsTests.cpp.o.d -o CMakeFiles/runtests.dir/LegacyOpsTests.cpp.o -c /home/zangetsu/proj/public/deeplearning4j/libnd4j/tests_cpu/layers_tests/LegacyOpsTests.cpp
In file included from /home/zangetsu/proj/public/deeplearning4j/libnd4j/tests_cpu/layers_tests/testlayers.h:34,
                 from /home/zangetsu/proj/public/deeplearning4j/libnd4j/tests_cpu/layers_tests/LegacyOpsTests.cpp:35:
/home/zangetsu/proj/public/deeplearning4j/libnd4j/include/legacy/NativeOps.h: In function ‘const char* getNpyArrayNameFromMap(void*, int, char*)’:
/home/zangetsu/proj/public/deeplearning4j/libnd4j/include/legacy/NativeOps.h:1285:1: warning: no return statement in function returning non-void [-Wreturn-type]
 1285 | }
      | ^
/home/zangetsu/proj/public/deeplearning4j/libnd4j/tests_cpu/layers_tests/LegacyOpsTests.cpp: In member function ‘virtual void LegacyOpsTests_reduce3_1_Test::TestBody()’:
/home/zangetsu/proj/public/deeplearning4j/libnd4j/tests_cpu/layers_tests/LegacyOpsTests.cpp:379:56: error: no matching function for call to ‘sd::ShapeBuilders::createShapeInfo(sd::DataType, char, int, sd::LongType [2])’
  379 |   auto shapeBuffer = sd::ShapeBuilders::createShapeInfo(sd::DataType::FLOAT32, 'c', 2, yShape);
      |                      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /home/zangetsu/proj/public/deeplearning4j/libnd4j/include/array/NDArray.h:34,
                 from /home/zangetsu/proj/public/deeplearning4j/libnd4j/tests_cpu/layers_tests/LegacyOpsTests.cpp:22:
/home/zangetsu/proj/public/deeplearning4j/libnd4j/include/helpers/ShapeBuilders.h:49:20: note: candidate: ‘static sd::LongType* sd::ShapeBuilders::createShapeInfo(sd::DataType, char, int, const LongType*, sd::memory::Workspace*, bool)’
   49 |   static LongType* createShapeInfo(const DataType dataType, const char order, int rank,
      |                    ^~~~~~~~~~~~~~~
/home/zangetsu/proj/public/deeplearning4j/libnd4j/include/helpers/ShapeBuilders.h:49:20: note:   candidate expects 6 arguments, 4 provided
/home/zangetsu/proj/public/deeplearning4j/libnd4j/include/helpers/ShapeBuilders.h:51:20: note: candidate: ‘static sd::LongType* sd::ShapeBuilders::createShapeInfo(sd::DataType, char, const std::vector<long long int>&, sd::memory::Workspace*)’
   51 |   static LongType* createShapeInfo(const DataType dataType, const char order,
      |                    ^~~~~~~~~~~~~~~
/home/zangetsu/proj/public/deeplearning4j/libnd4j/include/helpers/ShapeBuilders.h:52:69: note:   no known conversion for argument 3 from ‘int’ to ‘const std::vector<long long int>&’
   52 |                                        const std::vector<LongType>& shapeOnly,
      |                                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~
/home/zangetsu/proj/public/deeplearning4j/libnd4j/include/helpers/ShapeBuilders.h:54:20: note: candidate: ‘static sd::LongType* sd::ShapeBuilders::createShapeInfo(sd::DataType, char, const std::initializer_list<long long int>&, sd::memory::Workspace*)’
   54 |   static LongType* createShapeInfo(const DataType dataType, const char order,
      |                    ^~~~~~~~~~~~~~~
/home/zangetsu/proj/public/deeplearning4j/libnd4j/include/helpers/ShapeBuilders.h:55:79: note:   no known conversion for argument 3 from ‘int’ to ‘const std::initializer_list<long long int>&’
   55 |                                        const std::initializer_list<LongType>& shapeOnly,
      |                                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~
/home/zangetsu/proj/public/deeplearning4j/libnd4j/tests_cpu/layers_tests/LegacyOpsTests.cpp:380:57: error: no matching function for call to ‘sd::ShapeBuilders::createShapeInfo(sd::DataType, char, int, sd::LongType [1])’
  380 |   auto xShapeBuffer = sd::ShapeBuilders::createShapeInfo(sd::DataType::FLOAT32, 'c', 1, xShape);
      |                       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /home/zangetsu/proj/public/deeplearning4j/libnd4j/include/array/NDArray.h:34,
                 from /home/zangetsu/proj/public/deeplearning4j/libnd4j/tests_cpu/layers_tests/LegacyOpsTests.cpp:22:
/home/zangetsu/proj/public/deeplearning4j/libnd4j/include/helpers/ShapeBuilders.h:49:20: note: candidate: ‘static sd::LongType* sd::ShapeBuilders::createShapeInfo(sd::DataType, char, int, const LongType*, sd::memory::Workspace*, bool)’
   49 |   static LongType* createShapeInfo(const DataType dataType, const char order, int rank,
      |                    ^~~~~~~~~~~~~~~
/home/zangetsu/proj/public/deeplearning4j/libnd4j/include/helpers/ShapeBuilders.h:49:20: note:   candidate expects 6 arguments, 4 provided
/home/zangetsu/proj/public/deeplearning4j/libnd4j/include/helpers/ShapeBuilders.h:51:20: note: candidate: ‘static sd::LongType* sd::ShapeBuilders::createShapeInfo(sd::DataType, char, const std::vector<long long int>&, sd::memory::Workspace*)’
   51 |   static LongType* createShapeInfo(const DataType dataType, const char order,
      |                    ^~~~~~~~~~~~~~~
/home/zangetsu/proj/public/deeplearning4j/libnd4j/include/helpers/ShapeBuilders.h:52:69: note:   no known conversion for argument 3 from ‘int’ to ‘const std::vector<long long int>&’
   52 |                                        const std::vector<LongType>& shapeOnly,
      |                                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~
/home/zangetsu/proj/public/deeplearning4j/libnd4j/include/helpers/ShapeBuilders.h:54:20: note: candidate: ‘static sd::LongType* sd::ShapeBuilders::createShapeInfo(sd::DataType, char, const std::initializer_list<long long int>&, sd::memory::Workspace*)’
   54 |   static LongType* createShapeInfo(const DataType dataType, const char order,
      |                    ^~~~~~~~~~~~~~~
/home/zangetsu/proj/public/deeplearning4j/libnd4j/include/helpers/ShapeBuilders.h:55:79: note:   no known conversion for argument 3 from ‘int’ to ‘const std::initializer_list<long long int>&’
   55 |                                        const std::initializer_list<LongType>& shapeOnly,
      |                                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~
/home/zangetsu/proj/public/deeplearning4j/libnd4j/tests_cpu/layers_tests/LegacyOpsTests.cpp:389:12: error: type ‘<type error>’ argument given to ‘delete’, expected pointer
  389 |   delete[] shapeBuffer;
      |            ^~~~~~~~~~~
compilation terminated due to -fmax-errors=3.
make[2]: *** [tests_cpu/layers_tests/CMakeFiles/runtests.dir/build.make:762: tests_cpu/layers_tests/CMakeFiles/runtests.dir/LegacyOpsTests.cpp.o] Error 1
make[2]: Leaving directory '/home/zangetsu/proj/public/deeplearning4j/libnd4j/blasbuild/cpu'
make[1]: *** [CMakeFiles/Makefile2:542: tests_cpu/layers_tests/CMakeFiles/runtests.dir/all] Error 2
make[1]: Leaving directory '/home/zangetsu/proj/public/deeplearning4j/libnd4j/blasbuild/cpu'
make: *** [Makefile:166: all] Error 2
```

I have installed the following CUDA artifacts:
```
 zangetsu  X10SRA  …/public/deeplearning4j  cuda-latest-fixed  ls /home/zangetsu/devel/sdk/cuda-12/
drwxrwxr-x 2 zangetsu zangetsu 4,0K 2023-12-28 20:26 .
drwxrwxr-x 5 zangetsu zangetsu 4,0K 2023-12-28 20:45 ..
-rw-rw-r-- 1 zangetsu zangetsu 4,3K 2022-04-22 11:14 cuda-keyring_1.0-1_all.deb
-rw-rw-r-- 1 zangetsu zangetsu 3,1G 2023-11-09 07:53 cuda-repo-ubuntu2204-12-3-local_12.3.1-545.23.08-1_amd64.deb
-rw-rw-r-- 1 zangetsu zangetsu 846M 2023-12-28 19:34 cudnn-local-repo-ubuntu2204-8.9.7.29_1.0-1_amd64.deb
-rw-rw-r-- 1 zangetsu zangetsu 192M 2023-12-28 19:56 nccl-local-repo-ubuntu2204-2.19.3-cuda12.3_1.0-1_amd64.deb
-rw-rw-r-- 1 zangetsu zangetsu  22M 2023-12-04 19:28 nvcomp_3.0.5_x86_64_12.x.tgz

sudo apt-get -y install cuda-toolkit-12-3
sudo apt install libcudnn8=8.9.7.29-1+cuda12.2 libcudnn8-dev=8.9.7.29-1+cuda12.2 libcudnn8-samples=8.9.7.29-1+cuda12.2
sudo apt install libnccl2=2.19.3-1+cuda12.3 libnccl-dev=2.19.3-1+cuda12.3
sudo tar -xvf nvcomp_*.tgz -C /usr/local/cuda/include/ --strip-components=1 include/
sudo tar -xvf nvcomp_*.tgz -C /usr/local/cuda/lib64/ --strip-components=1 lib/
```

agibsonccc commented:
@archenroot just build skipping the c++ tests for now. All of the java side tests now run under platform-tests. I'm not sure what you're trying to do but you can do cpu with: 
mvn -Pcpu clean install -DskipTests -Dlibnd4j.buildthreadss=8
cuda with:
mvn -Pcuda -Dlibnd4j.chip=cuda -Dlibnd4j.buildthreads=8 clean install -DskipTests

That should build either cpu or cuda for you. Happy to fix anything outside of that. Thanks for taking a look at this!

archenroot commented:
@agibsonccc - thx for feedback.

I tested the following and both works:
```
mvn -Pcpu clean install -DskipTests -Dlibnd4j.buildthreadss=8
mvn -T 88 -Pgpu clean install -DskipTests -Dlibnd4j.buildthreadss=32
```

agibsonccc commented:
Great! Then yeah go ahead and use this branch for now. As you can see there's quite a few new changes coming in. I plan on splitting this up in to several cleaner PRS later.

Kali-Zoidberg commented:
Hey Adam, I'm having another issue compiling libnd4j and I think it's orginating at NativeOpExecutioner.cu. I've fixed the THROW_EXECUTION quote syntax error and missing semi colons but couldn't get it to compile (I'll leave a comment on the PR of the affected area). Would you be able to provide some insight into these error messages?

The error below originated building with CUDA 11.6 on Windows 10 w/ compute 8.0, MSYS2, c++17

I am currently trying to rebuild with CUDA 12.1 and reupdate if the build succeeds.

```[828/828] C:\WINDOWS\system32\cmd.exe /C "cd . && C:\msys64\mingw64\bin\cmake.exe -E vs_link_dll --intdir=blas\CMakeFiles\nd4jcuda.dir --rc=C:\PROGRA~2\WI3CF2~1\10\bin\100183~1.0\x64\rc.exe --mt=C:\PROGRA~2\WI3CF2~1\10\bin\100183~1.0\x64\mt.exe --manifests  -- C:\PROGRA~2\MICROS~2\2017\PROFES~1\VC\Tools\MSVC\1416~1.270\bin\Hostx64\x64\link.exe /nologo @CMakeFiles\nd4jcuda.rsp  /out:blas\nd4jcuda.dll /implib:blas\nd4jcuda.lib /pdb:blas\nd4jcuda.pdb /dll /version:0.0 /machine:x64 /INCREMENTAL:NO  && cd ."
FAILED: blas/nd4jcuda.dll blas/nd4jcuda.lib
C:\WINDOWS\system32\cmd.exe /C "cd . && C:\msys64\mingw64\bin\cmake.exe -E vs_link_dll --intdir=blas\CMakeFiles\nd4jcuda.dir --rc=C:\PROGRA~2\WI3CF2~1\10\bin\100183~1.0\x64\rc.exe --mt=C:\PROGRA~2\WI3CF2~1\10\bin\100183~1.0\x64\mt.exe --manifests  -- C:\PROGRA~2\MICROS~2\2017\PROFES~1\VC\Tools\MSVC\1416~1.270\bin\Hostx64\x64\link.exe /nologo @CMakeFiles\nd4jcuda.rsp  /out:blas\nd4jcuda.dll /implib:blas\nd4jcuda.lib /pdb:blas\nd4jcuda.pdb /dll /version:0.0 /machine:x64 /INCREMENTAL:NO  && cd ."
LINK: command "C:\PROGRA~2\MICROS~2\2017\PROFES~1\VC\Tools\MSVC\1416~1.270\bin\Hostx64\x64\link.exe /nologo @CMakeFiles\nd4jcuda.rsp /out:blas\nd4jcuda.dll /implib:blas\nd4jcuda.lib /pdb:blas\nd4jcuda.pdb /dll /version:0.0 /machine:x64 /INCREMENTAL:NO /MANIFEST:EMBED,ID=2" failed (exit code 1120) with the following output:
   Creating library blas\nd4jcuda.lib and object blas\nd4jcuda.exp
NativeOpExecutioner.cu.obj : error LNK2019: unresolved external symbol "public: static void __cdecl functions::transform::TransformAny<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >,class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > >::executeTransformShaped(struct dim3,struct CUstream_st * *,int,void const *,__int64 const *,__int64,void *,void *,__int64 const *,__int64,__int64 *,void *,__int64 const *,__int64 const *)" (?executeTransformShaped@?$TransformAny@V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@V12@@transform@functions@@SAXUdim3@@PEAPEAUCUstream_st@@HPEBXPEB_J_JPEAX534PEA_J533@Z) referenced in function "public: static void __cdecl NativeOpExecutioner::execTransformAny(class sd::LaunchContext *,int,void const *,__int64 const *,void const *,__int64 const *,void *,__int64 const *,void *,__int64 const *,void *,__int64 const *,__int64 const *,bool)" (?execTransformAny@NativeOpExecutioner@@SAXPEAVLaunchContext@sd@@HPEBXPEB_J12PEAX232322_N@Z)
NativeOpExecutioner.cu.obj : error LNK2019: unresolved external symbol "public: static void __cdecl functions::transform::TransformAny<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >,class std::basic_string<char16_t,struct std::char_traits<char16_t>,class std::allocator<char16_t> > >::executeTransformShaped(struct dim3,struct CUstream_st * *,int,void const *,__int64 const *,__int64,void *,void *,__int64 const *,__int64,__int64 *,void *,__int64 const *,__int64 const *)" (?executeTransformShaped@?$TransformAny@V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@V?$basic_string@_SU?$char_traits@_S@std@@V?$allocator@_S@2@@2@@transform@functions@@SAXUdim3@@PEAPEAUCUstream_st@@HPEBXPEB_J_JPEAX534PEA_J533@Z) referenced in function "public: static void __cdecl NativeOpExecutioner::execTransformAny(class sd::LaunchContext *,int,void const *,__int64 const *,void const *,__int64 const *,void *,__int64 const *,void *,__int64 const *,void *,__int64 const *,__int64 const *,bool)" (?execTransformAny@NativeOpExecutioner@@SAXPEAVLaunchContext@sd@@HPEBXPEB_J12PEAX232322_N@Z)
NativeOpExecutioner.cu.obj : error LNK2019: unresolved external symbol "public: static void __cdecl functions::transform::TransformAny<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >,class std::basic_string<char32_t,struct std::char_traits<char32_t>,class std::allocator<char32_t> > >::executeTransformShaped(struct dim3,struct CUstream_st * *,int,void const *,__int64 const *,__int64,void *,void *,__int64 const *,__int64,__int64 *,void *,__int64 const *,__int64 const *)" (?executeTransformShaped@?$TransformAny@V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@V?$basic_string@_UU?$char_traits@_U@std@@V?$allocator@_U@2@@2@@transform@functions@@SAXUdim3@@PEAPEAUCUstream_st@@HPEBXPEB_J_JPEAX534PEA_J533@Z) referenced in function "public: static void __cdecl NativeOpExecutioner::execTransformAny(class sd::LaunchContext *,int,void const *,__int64 const *,void const *,__int64 const *,void *,__int64 const *,void *,__int64 const *,void *,__int64 const *,__int64 const *,bool)" (?execTransformAny@NativeOpExecutioner@@SAXPEAVLaunchContext@sd@@HPEBXPEB_J12PEAX232322_N@Z)
NativeOpExecutioner.cu.obj : error LNK2019: unresolved external symbol "public: static void __cdecl functions::transform::TransformAny<class std::basic_string<char16_t,struct std::char_traits<char16_t>,class std::allocator<char16_t> >,class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > >::executeTransformShaped(struct dim3,struct CUstream_st * *,int,void const *,__int64 const *,__int64,void *,void *,__int64 const *,__int64,__int64 *,void *,__int64 const *,__int64 const *)" (?executeTransformShaped@?$TransformAny@V?$basic_string@_SU?$char_traits@_S@std@@V?$allocator@_S@2@@std@@V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@2@@transform@functions@@SAXUdim3@@PEAPEAUCUstream_st@@HPEBXPEB_J_JPEAX534PEA_J533@Z) referenced in function "public: static void __cdecl NativeOpExecutioner::execTransformAny(class sd::LaunchContext *,int,void const *,__int64 const *,void const *,__int64 const *,void *,__int64 const *,void *,__int64 const *,void *,__int64 const *,__int64 const *,bool)" (?execTransformAny@NativeOpExecutioner@@SAXPEAVLaunchContext@sd@@HPEBXPEB_J12PEAX232322_N@Z)
NativeOpExecutioner.cu.obj : error LNK2019: unresolved external symbol "public: static void __cdecl functions::transform::TransformAny<class std::basic_string<char16_t,struct std::char_traits<char16_t>,class std::allocator<char16_t> >,class std::basic_string<char16_t,struct std::char_traits<char16_t>,class std::allocator<char16_t> > >::executeTransformShaped(struct dim3,struct CUstream_st * *,int,void const *,__int64 const *,__int64,void *,void *,__int64 const *,__int64,__int64 *,void *,__int64 const *,__int64 const *)" (?executeTransformShaped@?$TransformAny@V?$basic_string@_SU?$char_traits@_S@std@@V?$allocator@_S@2@@std@@V12@@transform@functions@@SAXUdim3@@PEAPEAUCUstream_st@@HPEBXPEB_J_JPEAX534PEA_J533@Z) referenced in function "public: static void __cdecl NativeOpExecutioner::execTransformAny(class sd::LaunchContext *,int,void const *,__int64 const *,void const *,__int64 const *,void *,__int64 const *,void *,__int64 const *,void *,__int64 const *,__int64 const *,bool)" (?execTransformAny@NativeOpExecutioner@@SAXPEAVLaunchContext@sd@@HPEBXPEB_J12PEAX232322_N@Z)
NativeOpExecutioner.cu.obj : error LNK2019: unresolved external symbol "public: static void __cdecl functions::transform::TransformAny<class std::basic_string<char16_t,struct std::char_traits<char16_t>,class std::allocator<char16_t> >,class std::basic_string<char32_t,struct std::char_traits<char32_t>,class std::allocator<char32_t> > >::executeTransformShaped(struct dim3,struct CUstream_st * *,int,void const *,__int64 const *,__int64,void *,void *,__int64 const *,__int64,__int64 *,void *,__int64 const *,__int64 const *)" (?executeTransformShaped@?$TransformAny@V?$basic_string@_SU?$char_traits@_S@std@@V?$allocator@_S@2@@std@@V?$basic_string@_UU?$char_traits@_U@std@@V?$allocator@_U@2@@2@@transform@functions@@SAXUdim3@@PEAPEAUCUstream_st@@HPEBXPEB_J_JPEAX534PEA_J533@Z) referenced in function "public: static void __cdecl NativeOpExecutioner::execTransformAny(class sd::LaunchContext *,int,void const *,__int64 const *,void const *,__int64 const *,void *,__int64 const *,void *,__int64 const *,void *,__int64 const *,__int64 const *,bool)" (?execTransformAny@NativeOpExecutioner@@SAXPEAVLaunchContext@sd@@HPEBXPEB_J12PEAX232322_N@Z)
NativeOpExecutioner.cu.obj : error LNK2019: unresolved external symbol "public: static void __cdecl functions::transform::TransformAny<class std::basic_string<char32_t,struct std::char_traits<char32_t>,class std::allocator<char32_t> >,class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > >::executeTransformShaped(struct dim3,struct CUstream_st * *,int,void const *,__int64 const *,__int64,void *,void *,__int64 const *,__int64,__int64 *,void *,__int64 const *,__int64 const *)" (?executeTransformShaped@?$TransformAny@V?$basic_string@_UU?$char_traits@_U@std@@V?$allocator@_U@2@@std@@V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@2@@transform@functions@@SAXUdim3@@PEAPEAUCUstream_st@@HPEBXPEB_J_JPEAX534PEA_J533@Z) referenced in function "public: static void __cdecl NativeOpExecutioner::execTransformAny(class sd::LaunchContext *,int,void const *,__int64 const *,void const *,__int64 const *,void *,__int64 const *,void *,__int64 const *,void *,__int64 const *,__int64 const *,bool)" (?execTransformAny@NativeOpExecutioner@@SAXPEAVLaunchContext@sd@@HPEBXPEB_J12PEAX232322_N@Z)
NativeOpExecutioner.cu.obj : error LNK2019: unresolved external symbol "public: static void __cdecl functions::transform::TransformAny<class std::basic_string<char32_t,struct std::char_traits<char32_t>,class std::allocator<char32_t> >,class std::basic_string<char16_t,struct std::char_traits<char16_t>,class std::allocator<char16_t> > >::executeTransformShaped(struct dim3,struct CUstream_st * *,int,void const *,__int64 const *,__int64,void *,void *,__int64 const *,__int64,__int64 *,void *,__int64 const *,__int64 const *)" (?executeTransformShaped@?$TransformAny@V?$basic_string@_UU?$char_traits@_U@std@@V?$allocator@_U@2@@std@@V?$basic_string@_SU?$char_traits@_S@std@@V?$allocator@_S@2@@2@@transform@functions@@SAXUdim3@@PEAPEAUCUstream_st@@HPEBXPEB_J_JPEAX534PEA_J533@Z) referenced in function "public: static void __cdecl NativeOpExecutioner::execTransformAny(class sd::LaunchContext *,int,void const *,__int64 const *,void const *,__int64 const *,void *,__int64 const *,void *,__int64 const *,void *,__int64 const *,__int64 const *,bool)" (?execTransformAny@NativeOpExecutioner@@SAXPEAVLaunchContext@sd@@HPEBXPEB_J12PEAX232322_N@Z)
NativeOpExecutioner.cu.obj : error LNK2019: unresolved external symbol "public: static void __cdecl functions::transform::TransformAny<class std::basic_string<char32_t,struct std::char_traits<char32_t>,class std::allocator<char32_t> >,class std::basic_string<char32_t,struct std::char_traits<char32_t>,class std::allocator<char32_t> > >::executeTransformShaped(struct dim3,struct CUstream_st * *,int,void const *,__int64 const *,__int64,void *,void *,__int64 const *,__int64,__int64 *,void *,__int64 const *,__int64 const *)" (?executeTransformShaped@?$TransformAny@V?$basic_string@_UU?$char_traits@_U@std@@V?$allocator@_U@2@@std@@V12@@transform@functions@@SAXUdim3@@PEAPEAUCUstream_st@@HPEBXPEB_J_JPEAX534PEA_J533@Z) referenced in function "public: static void __cdecl NativeOpExecutioner::execTransformAny(class sd::LaunchContext *,int,void const *,__int64 const *,void const *,__int64 const *,void *,__int64 const *,void *,__int64 const *,void *,__int64 const *,__int64 const *,bool)" (?execTransformAny@NativeOpExecutioner@@SAXPEAVLaunchContext@sd@@HPEBXPEB_J12PEAX232322_N@Z)
NativeOpExecutioner.cu.obj : error LNK2019: unresolved external symbol "public: static void __cdecl functions::scalar::ScalarTransform<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >,class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >,class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > >::executeCudaAlongDimension(struct dim3 &,struct CUstream_st * *,int,void const *,__int64 const *,void *,__int64 const *,void const *,void *,__int64 *,__int64,__int64 const *,__int64 const *,__int64 const *,__int64 const *)" (?executeCudaAlongDimension@?$ScalarTransform@V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@V12@V12@@scalar@functions@@SAXAEAUdim3@@PEAPEAUCUstream_st@@HPEBXPEB_JPEAX324PEA_J_J3333@Z) referenced in function "public: static void __cdecl NativeOpExecutioner::execScalar(class sd::LaunchContext *,int,void const *,__int64 const *,void const *,__int64 const *,void *,void *,__int64 const *,void *,__int64 const *,void const *,__int64 const *,void const *,__int64 const *,__int64 *,__int64,__int64 const *,__int64 const *,__int64 const *,__int64 const *)" (?execScalar@NativeOpExecutioner@@SAXPEAVLaunchContext@sd@@HPEBXPEB_J12PEAX32321212PEA_J_J2222@Z)
NativeOpExecutioner.cu.obj : error LNK2019: unresolved external symbol "public: static void __cdecl functions::scalar::ScalarTransform<class std::basic_string<char16_t,struct std::char_traits<char16_t>,class std::allocator<char16_t> >,class std::basic_string<char16_t,struct std::char_traits<char16_t>,class std::allocator<char16_t> >,class std::basic_string<char16_t,struct std::char_traits<char16_t>,class std::allocator<char16_t> > >::executeCudaAlongDimension(struct dim3 &,struct CUstream_st * *,int,void const *,__int64 const *,void *,__int64 const *,void const *,void *,__int64 *,__int64,__int64 const *,__int64 const *,__int64 const *,__int64 const *)" (?executeCudaAlongDimension@?$ScalarTransform@V?$basic_string@_SU?$char_traits@_S@std@@V?$allocator@_S@2@@std@@V12@V12@@scalar@functions@@SAXAEAUdim3@@PEAPEAUCUstream_st@@HPEBXPEB_JPEAX324PEA_J_J3333@Z) referenced in function "public: static void __cdecl NativeOpExecutioner::execScalar(class sd::LaunchContext *,int,void const *,__int64 const *,void const *,__int64 const *,void *,void *,__int64 const *,void *,__int64 const *,void const *,__int64 const *,void const *,__int64 const *,__int64 *,__int64,__int64 const *,__int64 const *,__int64 const *,__int64 const *)" (?execScalar@NativeOpExecutioner@@SAXPEAVLaunchContext@sd@@HPEBXPEB_J12PEAX32321212PEA_J_J2222@Z)
NativeOpExecutioner.cu.obj : error LNK2019: unresolved external symbol "public: static void __cdecl functions::scalar::ScalarTransform<class std::basic_string<char32_t,struct std::char_traits<char32_t>,class std::allocator<char32_t> >,class std::basic_string<char32_t,struct std::char_traits<char32_t>,class std::allocator<char32_t> >,class std::basic_string<char32_t,struct std::char_traits<char32_t>,class std::allocator<char32_t> > >::executeCudaAlongDimension(struct dim3 &,struct CUstream_st * *,int,void const *,__int64 const *,void *,__int64 const *,void const *,void *,__int64 *,__int64,__int64 const *,__int64 const *,__int64 const *,__int64 const *)" (?executeCudaAlongDimension@?$ScalarTransform@V?$basic_string@_UU?$char_traits@_U@std@@V?$allocator@_U@2@@std@@V12@V12@@scalar@functions@@SAXAEAUdim3@@PEAPEAUCUstream_st@@HPEBXPEB_JPEAX324PEA_J_J3333@Z) referenced in function "public: static void __cdecl NativeOpExecutioner::execScalar(class sd::LaunchContext *,int,void const *,__int64 const *,void const *,__int64 const *,void *,void *,__int64 const *,void *,__int64 const *,void const *,__int64 const *,void const *,__int64 const *,__int64 *,__int64,__int64 const *,__int64 const *,__int64 const *,__int64 const *)" (?execScalar@NativeOpExecutioner@@SAXPEAVLaunchContext@sd@@HPEBXPEB_J12PEAX32321212PEA_J_J2222@Z)
StringUtils.cpp.obj : error LNK2019: unresolved external symbol "public: void __cdecl sd::NDArray::p<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > >(__int64,class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >)" (??$p@V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@NDArray@sd@@QEAAX_JV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@Z) referenced in function "public: static void __cdecl sd::StringUtils::setValueForDifferentDataType(class sd::NDArray *,__int64,class sd::NDArray *,enum sd::DataType)" (?setValueForDifferentDataType@StringUtils@sd@@SAXPEAVNDArray@2@_J0W4DataType@2@@Z)
StringUtils.cpp.obj : error LNK2019: unresolved external symbol "public: void __cdecl sd::NDArray::p<class std::basic_string<char16_t,struct std::char_traits<char16_t>,class std::allocator<char16_t> > >(__int64,class std::basic_string<char16_t,struct std::char_traits<char16_t>,class std::allocator<char16_t> >)" (??$p@V?$basic_string@_SU?$char_traits@_S@std@@V?$allocator@_S@2@@std@@@NDArray@sd@@QEAAX_JV?$basic_string@_SU?$char_traits@_S@std@@V?$allocator@_S@2@@std@@@Z) referenced in function "public: static void __cdecl sd::StringUtils::setValueForDifferentDataType(class sd::NDArray *,__int64,class sd::NDArray *,enum sd::DataType)" (?setValueForDifferentDataType@StringUtils@sd@@SAXPEAVNDArray@2@_J0W4DataType@2@@Z)
StringUtils.cpp.obj : error LNK2019: unresolved external symbol "public: void __cdecl sd::NDArray::p<class std::basic_string<char32_t,struct std::char_traits<char32_t>,class std::allocator<char32_t> > >(__int64,class std::basic_string<char32_t,struct std::char_traits<char32_t>,class std::allocator<char32_t> >)" (??$p@V?$basic_string@_UU?$char_traits@_U@std@@V?$allocator@_U@2@@std@@@NDArray@sd@@QEAAX_JV?$basic_string@_UU?$char_traits@_U@std@@V?$allocator@_U@2@@std@@@Z) referenced in function "public: static void __cdecl sd::StringUtils::setValueForDifferentDataType(class sd::NDArray *,__int64,class sd::NDArray *,enum sd::DataType)" (?setValueForDifferentDataType@StringUtils@sd@@SAXPEAVNDArray@2@_J0W4DataType@2@@Z)
NDArray.cu.obj : error LNK2019: unresolved external symbol "public: static class sd::NDArray __cdecl sd::NDArrayFactory::create<class std::vector<__int64,class std::allocator<__int64> > >(enum sd::DataType,class std::vector<__int64,class std::allocator<__int64> >,class sd::LaunchContext *)" (??$create@V?$vector@_JV?$allocator@_J@std@@@std@@@NDArrayFactory@sd@@SA?AVNDArray@1@W4DataType@1@V?$vector@_JV?$allocator@_J@std@@@std@@PEAVLaunchContext@1@@Z) referenced in function "public: class sd::NDArray __cdecl sd::NDArray::broadcastTo(class std::vector<__int64,class std::allocator<__int64> > const &)" (?broadcastTo@NDArray@sd@@QEAA?AV12@AEBV?$vector@_JV?$allocator@_J@std@@@std@@@Z)
NDArray.cu.obj : error LNK2019: unresolved external symbol "public: static class std::vector<__int64,class std::allocator<__int64> > __cdecl sd::StringUtils::calculateOffsetsForTargetDataType<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > >(class sd::NDArray const *)" (??$calculateOffsetsForTargetDataType@V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@StringUtils@sd@@SA?AV?$vector@_JV?$allocator@_J@std@@@std@@PEBVNDArray@1@@Z) referenced in function "public: class sd::NDArray __cdecl sd::NDArray::asS<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > >(void)const " (??$asS@V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@NDArray@sd@@QEBA?AV01@XZ)
NDArray.cu.obj : error LNK2019: unresolved external symbol "public: static void __cdecl sd::StringUtils::convertStringsForDifferentDataType<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > >(class sd::NDArray const *,class sd::NDArray *)" (??$convertStringsForDifferentDataType@V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@StringUtils@sd@@SAXPEBVNDArray@1@PEAV21@@Z) referenced in function "public: class sd::NDArray __cdecl sd::NDArray::asS<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > >(void)const " (??$asS@V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@NDArray@sd@@QEBA?AV01@XZ)
NDArray.cu.obj : error LNK2019: unresolved external symbol "public: static class std::vector<__int64,class std::allocator<__int64> > __cdecl sd::StringUtils::calculateOffsetsForTargetDataType<class std::basic_string<char16_t,struct std::char_traits<char16_t>,class std::allocator<char16_t> > >(class sd::NDArray const *)" (??$calculateOffsetsForTargetDataType@V?$basic_string@_SU?$char_traits@_S@std@@V?$allocator@_S@2@@std@@@StringUtils@sd@@SA?AV?$vector@_JV?$allocator@_J@std@@@std@@PEBVNDArray@1@@Z) referenced in function "public: class sd::NDArray __cdecl sd::NDArray::asS<class std::basic_string<char16_t,struct std::char_traits<char16_t>,class std::allocator<char16_t> > >(void)const " (??$asS@V?$basic_string@_SU?$char_traits@_S@std@@V?$allocator@_S@2@@std@@@NDArray@sd@@QEBA?AV01@XZ)
NDArray.cu.obj : error LNK2019: unresolved external symbol "public: static void __cdecl sd::StringUtils::convertStringsForDifferentDataType<class std::basic_string<char16_t,struct std::char_traits<char16_t>,class std::allocator<char16_t> > >(class sd::NDArray const *,class sd::NDArray *)" (??$convertStringsForDifferentDataType@V?$basic_string@_SU?$char_traits@_S@std@@V?$allocator@_S@2@@std@@@StringUtils@sd@@SAXPEBVNDArray@1@PEAV21@@Z) referenced in function "public: class sd::NDArray __cdecl sd::NDArray::asS<class std::basic_string<char16_t,struct std::char_traits<char16_t>,class std::allocator<char16_t> > >(void)const " (??$asS@V?$basic_string@_SU?$char_traits@_S@std@@V?$allocator@_S@2@@std@@@NDArray@sd@@QEBA?AV01@XZ)
NDArray.cu.obj : error LNK2019: unresolved external symbol "public: static class std::vector<__int64,class std::allocator<__int64> > __cdecl sd::StringUtils::calculateOffsetsForTargetDataType<class std::basic_string<char32_t,struct std::char_traits<char32_t>,class std::allocator<char32_t> > >(class sd::NDArray const *)" (??$calculateOffsetsForTargetDataType@V?$basic_string@_UU?$char_traits@_U@std@@V?$allocator@_U@2@@std@@@StringUtils@sd@@SA?AV?$vector@_JV?$allocator@_J@std@@@std@@PEBVNDArray@1@@Z) referenced in function "public: class sd::NDArray __cdecl sd::NDArray::asS<class std::basic_string<char32_t,struct std::char_traits<char32_t>,class std::allocator<char32_t> > >(void)const " (??$asS@V?$basic_string@_UU?$char_traits@_U@std@@V?$allocator@_U@2@@std@@@NDArray@sd@@QEBA?AV01@XZ)
NDArray.cu.obj : error LNK2019: unresolved external symbol "public: static void __cdecl sd::StringUtils::convertStringsForDifferentDataType<class std::basic_string<char32_t,struct std::char_traits<char32_t>,class std::allocator<char32_t> > >(class sd::NDArray const *,class sd::NDArray *)" (??$convertStringsForDifferentDataType@V?$basic_string@_UU?$char_traits@_U@std@@V?$allocator@_U@2@@std@@@StringUtils@sd@@SAXPEBVNDArray@1@PEAV21@@Z) referenced in function "public: class sd::NDArray __cdecl sd::NDArray::asS<class std::basic_string<char32_t,struct std::char_traits<char32_t>,class std::allocator<char32_t> > >(void)const " (??$asS@V?$basic_string@_UU?$char_traits@_U@std@@V?$allocator@_U@2@@std@@@NDArray@sd@@QEBA?AV01@XZ)
blas\nd4jcuda.dll : fatal error LNK1120: 22 unresolved externals
ninja: build stopped: subcommand failed.
```

Kali-Zoidberg commented on libnd4j/include/legacy/cuda/NativeOpExecutioner.cu:
```diff
@@ -118,125 +111,155 @@ void NativeOpExecutioner::execPairwiseTransform(sd::LaunchContext* lc, int opNum
+  if (DataTypeUtils::isS(xType) || DataTypeUtils::isS(zType)) {
+    THROW_EXCEPTION(
+        "NativeOpExecutioner::execSummaryStatsScalar:: unable to execute on strings. Please write logic higher level "
+        "in each op for the string data type.")
```
Quotes w/o concatenation and a missing semi colon. This appears a couple more times in this file.

agibsonccc replied:
Ah great catch! Let me push up a fix for this. Thanks! I'm currently in the middle of something else but should be able to get to this really fast in a commit.

archenroot commented:
This is mega PR :-)

agibsonccc commented:
Thanks for checking folks! I'll be breaking this up in the next week or so

agibsonccc committed:
Add new ndarray logging to track changes to arrays over time.
Adds array registry for tracking arrays in a central location mapped by id.
Adds event instrumentation to different parts of the nd4j api to track data at different parts of an array's lifecyle.

agibsonccc committed:
Fix bidirectional tests

agibsonccc committed:
Clean up print statements, delete unused classes.

agibsonccc committed:
Removes graveslstm, fix comp graph test for bidirectional

agibsonccc committed:
Fix lstm gradient check tests
Fix view offset detection by omitting views from element wise stride based offset calculations.

agibsonccc committed:
Fix up shape.h linking
Add new check input change which detects when an op changes an input and throws an exception when this is the case.

agibsonccc committed:
Add c++ based log ndarray events (print statements in ndarray and databuffer) to detect subtle move/copy bugs

agibsonccc committed:
Fix subtle view issues
Add explicit destructor for ndarrays
Improve logging for ndarray constructors with separate environment variable

agibsonccc committed:
Fix simplernn gemm

agibsonccc committed:
Add missing function

agibsonccc committed:
Fix conv3d gradient checks

agibsonccc committed:
Fix convolution output size calculations for 2/3d

agibsonccc committed:
Rewrite locallyconnected 2d.
Update address sanitizer build to link correctly.
Revert op.z() caching due to eager mode.
Remove VEDA/VEDNN
Convert all smart pointer databuffers to normal pointers.
Due to external buffers they produce very inconnsistent results.

agibsonccc committed:
WIP: update conv2d to work with different weight formats.

agibsonccc committed:
Finish conv2d layer conversion

agibsonccc committed:
Fix conv2d gradient checks (ports java logic to c++)
Rewrites conv2d to delegate to c++ while keeping the same logic.

agibsonccc committed:
Remove print statements
Fix bias conv2d gradient

agibsonccc committed:
Fix nchw setting in conv2d.
Fix sizes for locally connected 2d.

agibsonccc committed:
Clean up print statements
Fix NHWC case for new conv2d invocation.

agibsonccc committed:
Add creation stack trace tracking for constructors
when functrace is turned on.
Fix delete in constant shape helper when a shape descirptor exists.

agibsonccc committed:
Fix batched gemm indexing.
Delete more aurora

agibsonccc committed:
Fix constant shape helper with reshape in place.
Add validation for cache values that are being stored.

agibsonccc committed:
Fix more constant buffer cache in place modification.

agibsonccc committed:
Fix grad checks for conv2d bp.

agibsonccc committed:
Ensure compatibility with old dl4j conv2d implementation.

agibsonccc committed:
Get rid of tests.

agibsonccc committed:
Add difference detector. This allows detection of differences in ops to quickly find regressions across versions of dl4j.

agibsonccc committed:
Fix LSTM gradient checks/dup issue

agibsonccc committed:
Update readme, add missing licenses

agibsonccc committed:
Fix f order based indexOffset
Add PRINT_INDICES macro for debugging.

agibsonccc committed:
Fix up reporting for log analyzer
Fix address sanitizer warnings for new[]/delete[] vs new/delete.

agibsonccc committed:
WIP: Rewrite c++ conv2d to achieve parity with  dl4j implementation.

agibsonccc committed:
fix conv2d backprop with nhwc, fix dl4j level conv2d backprop
